### ISSUE TITLE
`ghcide` runs out of memory

### ISSUE THREAD

**Issue by ocharles** [digital-asset/ghcide#6]
Not sure what I can do to help with this, the codebase is proprietary, but I *may* be able to give you private access to it. My working directory has 1228 `.hs` files. `ghcide` (run from my shell) makes a lot of progress, and then gets killed as it runs out of memory (at approx. 6.2GB resident memory usage).

**Comment by cocreature** [comment]
We haven’t spend too much time on memory usage so far and our efforts have been mostly focused on testing it on DAML usage (although that should be roughly comparable) so sadly I’m not too surprised by high memory usage. Do you have some comparison of how much memory GHCi or maybe haskell-ide-engine uses on this project?

**Comment by ocharles** [comment]
I have never loaded the whole project into `ghci`, and am not entirely sure what the best way to test that wolud be. Should I try and load all known `.hs` files at the same time or something?

**Comment by ocharles** [comment]
I should note that this codebase is actually split into about 20 different projects. I'm trying to run `ghcide` from the top level directory (with `cabal.project`) with a direct-style cradle that would load all projects in a single session.

**Comment by cocreature** [comment]
You should be able to translate your direct-style cradle into a `.ghci` file that sets a bunch of `:set -i` paths and then `:l` the file you are trying to run `ghcide` on.

**Comment by ocharles** [comment]
Ah right, I'm not actually trying to run `ghcide` on any specific file - just doing the "did it work?" check of running `ghcide` at the command line with no arguments, which as I understand loads all files at once.

**Comment by cocreature** [comment]
Ah I see, in that case it recursively searches for `.hs` files and tries to load them. You can also invoke `ghcide` with a specific file path in which case it will try to load the transitive closure of that so I would try to choose a file high up in your dependency tree and compare that to ghci.

**Comment by cocreature** [comment]
Unless I am missing something max residency here is lower than the 2.8GB you get in GHCi? We can do a bit better than GHCi but I’m not sure there is that much we can do here. 1.5GB vs 2.8GB already seems like a decent improvement.

Leaving that aside, I’d be interested in where you need to make a difference for the `.ghci` file and the `hie-bios` cradle.

**Comment by ocharles** [comment]
Yes, but I don't believe `ghcide` is even doing the same thing as `ghci`, that's what I meant by:

> I don't trust it - my `hie.yaml` doesn't even have the same set of flags that I needed for `ghci`!

E.g., there are modules missing from the search path, but `ghcide` reports no error. Also, this ends up importing (somewhere) the same modules that cause #25, yet there is no error here. So I'm a bit suspicious about that max residency.

**Comment by cocreature** [comment]
`ghcide` has some logic (inherited from DAML, we should probably refactor this) where it tries to locate a module relative to the current one even if that’s not in the search path so if it resolves an import in `a/b/C.hs` it will implicitely add `a/b` to the search path for finding imports in `C.hs`. That could explain why you need less search paths in `ghcide`.

**Comment by ocharles** [comment]
Ah, that sounds plausible. The missing path is relative to the `Main.hs` I'm supplying. I'll play around with some other entry points and see what I can get.

**Comment by ndmitchell** [comment]
Probably worth trying again with the latest. It might have gotten better or worse, but that information is probably useful. Note there are further efforts to reduce the memory usage with #355 `.hi` file support.

**Comment by ocharles** [comment]
Thanks for the suggestion. On our work codebase, I still can't run `ghcide` (`HEAD`) from the command line without it getting terminated. It used 10GB memory in just under 4 minutes, at which point my system killed it due to running out of memory:

```
Jan 30 08:19:51 nixos earlyoom[680]: mem avail:  1664 of 15752 MiB (10 %), swap free:    0 of    0 MiB ( 0 %)
Jan 30 08:19:51 nixos earlyoom[680]: low memory! at or below SIGTERM limits: mem 10 %, swap 10 %
Jan 30 08:19:51 nixos earlyoom[680]: sending SIGTERM to process 15681 "ghcide": badness 602, VmRSS 9473 MiB
```

**Comment by pepeiborra** [comment]
Could you try with the branch at https://github.com/pepeiborra/ghcide/tree/interface-files-rebased-again and let us know in https://github.com/digital-asset/ghcide/issues/355 ? It's still a work in progress but I have seen significant space reductions in large codebases.

**Comment by ocharles** [comment]
@pepeiborra Gladly! Do I have to do anything special to take advantage of this work? E.g., compile `.hi`/`.hie` files and put them anywhere special? We use a direct cradle

**Comment by pepeiborra** [comment]
Yes. You need to use ghc 8.8, make sure to build the project with Cabal, and optionally tell ghc to generate .hie files. It’s all explained in the Readme

**Comment by ocharles** [comment]
Thanks, I was not aware the readme has changed. I will try now.

**Comment by ocharles** [comment]
I'm still a little unclear if I should put the `.hie` files anywhere special. As I use cabal, they will end up in `dist-newstyle`, but there is no reference to that from my cradle. Does that matter? Would you like me to take this discussion to #355?

**Comment by pepeiborra** [comment]
Fix it with -hiedir 
Hie files are only used for goto definition, so not terribly important if you are only interested in gauging the memory savings.

**Comment by pepeiborra** [comment]
Is this any better now?

**Comment by ocharles** [comment]
Yes, but only because I bought 32GB RAM :laughing: 

I need to try out the new interface files work, I'll see if I can upgrade soon.

**Comment by jneira** [comment]
@ocharles did you have the chance to test the new version?

**Comment by tek** [comment]
I also have a project that used to fill up 30G of ram in a matter of minutes, and with current master, it seems to have improved significantly. Saturating at about 5G, as far as I've observed.
